New browser RUM guide for Electron JS apps (#19193)
DataDog/documentation: content/en/real_user_monitoring/guide/monitor-electron-applications-using-browser-sdk.md

further_reading:
  - link: '/real_user_monitoring/'
    tag: 'Documentation'
    text: 'Learn about Real User Monitoring'
---

<div class="alert alert-warning">
    Electron support for RUM is in beta.
</div>

## Overview

[Electron][1] is an open source framework that can be used to build cross-platform macOS and Windows desktop applications.

You can install and configure the Datadog Browser SDK to start monitoring applications built using Electron JS.

**Note**: The Browser SDK only supports monitoring the *renderer processes* of an application. It does not initialize or monitor anything installed on the *main process*. For more information, see Electron's documentation on its [renderer process][3].

{{< img src="real_user_monitoring/guide/monitor-electron-applications-using-browser-sdk/diagram_electron-js-browser-rum.png" alt="The Datadog Browser SDK only supports monitoring the renderer processes of an Electron application" style="width:100%" >}}

## Installation

To install the Datadog Browser SDK to support Electron apps:

1. Set up and install [RUM Browser Monitoring][2] inside **every renderer process**, following the steps for CDN sync, CDN async, or npm.

> 2. Set the `allowFallbackToLocalStorage` parameter to true in the RUM initialization configuration of each renderer process, as shown below. 

Reviewer: ```suggestion
2. Set the `allowFallbackToLocalStorage` parameter to `true` in the RUM initialization configuration of each renderer process, as shown below. 
```

>    **Note**: This setting allows Datadog to collect RUM without relying on browser cookies.

Reviewer: This sentence reads a little oddly to me. Is there another word you can add after RUM? Perhaps "data," so the sentence would read something like "allows Datadog to collect RUM data without relying on..."

   
   - If you are targeting pages **available on the internet** (using the `https://` protocol), you **do not** need this parameter.
   - if you are embedding pages **inside your application** (using the `file://` protocol), Datadog needs to store sessions in local storage, as cookies are not available. 

   ```javascript
   datadogRum.init({
     applicationId: '<DATADOG_APPLICATION_ID>',
     clientToken: '<DATADOG_CLIENT_TOKEN>',
     site: '<DATADOG_SITE>',
     ...
     allowFallbackToLocalStorage: true
     });
   ```
 

> 3. Once you've configured the SDK correctly, you can see data in the [RUM Explorer][4].

Reviewer: ```suggestion
3. Once you've configured the SDK correctly, your data populates the [RUM Explorer][4].
```

## Troubleshooting

### Support for hybrid Electron applications
The same-origin policy prevents tracking an application for the same session in which pages load from both local (`file://`) and remote (`http(s)://`).

This means that an application that uses Electron to embed a landing page, then later redirects the user to a website hosted on the Internet results in two sessions being created for that user - one for the embedded local files (`file://`) landing part of the application, and one for the remote part (`https://` files available on the internet).

### Short-lived sessions for instances with multiple windows at once

> An issue with local storage replication latency between windows can cause a short-lived session to be created (<1 second). To workaround this, ensure multiple windows are created and initialized with a gap of more than 10 ms.

Reviewer: ```suggestion
An issue with local storage replication latency between windows can cause a short-lived session to be created (<1 second). To work around this, ensure multiple windows are created and initialized with a gap of more than 10 ms.
```
